Spatial.look_at_dir  (look at direction function)

kubecz3k opened the pull request:
Basically a syntax sugar, what this function does is using globalSpatialPos + lookDirection together with normal look_at() function

TheoXD commented:
I don't see any usecase for this, please elaborate. Look inside set_look_at() for a sec, v_z = p_eye - p_target where p_eye = get_global_transform().origin, and your p_target = (get_global_transform().origin + p_target)   . So origin will be canceled out, you can assume spatial is always at (0, 0, 0), where else will this work?

kubecz3k commented:
Hmm I'm using this right now and origin is not canceled out...

TheoXD commented:
Oh so it's your way to only pass in direction directly, Some unnecessary calculation going on but I guess it's ok  :)

kubecz3k commented:
@TheoXD Oh you mean it would be possible to use Matrix directly and omit `look_at` function call?

kubecz3k commented:
Hmm to omit those calculations I would probably need to implement new function in Transform, would prefer to not touch this

kubecz3k commented:
however it would be easy since what the function does at the beginning is to calculate direction vector (yeah @TheoXD now I understand what you meant).. Do you think it's necessary?

TheoXD commented:
In a long run might not be the wisest approach, but you can create a function next to Transform::looking_at() such as "looking_at_dir", then you pass a null instead of origin and do a null check. Plus accessible not only from Spatial but also from Transform.

kubecz3k commented:
Ok I will investigate this further :)

TheoXD commented:
Think that there are 3*2 float operations going on, plus normalize() when it's not needed (assuming direction is already normalized).

kubecz3k commented:
Updated Transform to have `set_look_at_dir` method.
I also renamed method `Spatial.look_in_dir` to `Spatial.look_at_dir` since Spatial that's standing at 0,0,0 and is looking in 0,0,1 direction will look at point 0,0,1 (and not at 0,0,-1 which is what some people who are more into opengl could be expecting with look_in name).
In the end I didn't skip normalization of direction vector since this would make this method more error prone.

kubecz3k committed:
Spatial.look_at_dir function

Adds new method set_look_at_dir to transform and links it with Spatial

reduz commented:
I still don't understand the difference between this and regular look_at() for Spatial

kubecz3k commented on core/math/transform.cpp:
```diff
@@ -113,6 +113,41 @@ void Transform::set_look_at( const Vector3& p_eye, const Vector3& p_target, cons
+	/* Make rotation matrix */
+
+	/* Z vector */
+	v_z = -p_target;
```
This is inconsistent,  with the opengl where -z is positive, from the perspective of time I'm not sure if this is good

kubecz3k commented:
@reduz well most of the time you want to rotate your model to face not a point but a direction. Mean, when you are firing you want your bullet to face moving direction. When enemy is moving he usually is facing movement direction, when player is using analog to steer character, you also have a direction, and it  can be easily interpolated. 
If it's about tech difference well, the main difference is look_at() is always working in global space coordinates, while look_at_dir is not. 
However when I'm thinking about this pr I'm not sure if negating look direction was a good idea in the first place - wanted to allow user to use direction in z positive coord system, but now I'm wondering if it was a good idea...  https://github.com/godotengine/godot/pull/5725/files#diff-94df3c579a1bbbb928fdb97413222215R130
If you don't think the pr is not useful I'm fine and it can be closed.

tagcup commented:
> well most of the time you want to rotate your model to face not a point but a direction. Mean, when you are firing you want your bullet to face moving direction. 

But that's look_at does to the Transform already. It changes the orientation towards the direction `p_target - p_eye`, and ensures that you're located at `p_eye`.

If you don't want to move your character, and want to face to a particular direction given by a vector `direction`, you simply need to do `look_at(get_transform().origin, get_transform().origin + direction, up)` where `get_transform().origin + direction` is clearly your target. I mean it's just a matter of rearranging the simple equation `direction = p_target - p_eye`.

kubecz3k commented:
I'm closing this pr since as far as I know in 3.0 there is good chance there will be local_look_at()
